pdr_backend/ppss/abstract/base_ss.py

```diff
@@ -0,0 +1,158 @@
+from abc import ABC
+from typing import Dict, List, Optional, Set, Tuple, Union
+
+from enforce_typing import enforce_types
+
+from pdr_backend.cli.arg_feed import ArgFeed, ArgFeeds
+from pdr_backend.cli.arg_pair import ArgPair
+from pdr_backend.subgraph.subgraph_feed import SubgraphFeed
+
+
+class MultiFeedSS(ABC):
+    FEEDS_KEY = ""
+
+    @enforce_types
+    def __init__(self, d: dict, assert_feed_attributes: Optional[List] = None):
+        assert self.__class__.FEEDS_KEY
+        self.d = d
+        # save self.exchs_dict
+        self.exchs_dict: dict = {}  # e.g. {"binance" : ccxt.binance()}
+        feeds = ArgFeeds.from_strs(self.input_feeds_strs)
+        for feed in feeds:
+            exchange_class = feed.exchange.exchange_class
+            self.exchs_dict[str(feed.exchange)] = exchange_class()
+            if assert_feed_attributes:
+                for attr in assert_feed_attributes:
+                    assert getattr(feed, attr)
+
+    # --------------------------------
+    # yaml properties
+    @property
+    def input_feeds_strs(self) -> List[str]:
+        return self.d[self.__class__.FEEDS_KEY]  # eg ["binance BTC/USDT ohlcv",..]
+
+    # --------------------------------
+
+    @property
+    def n_exchs(self) -> int:
+        return len(self.exchs_dict)
+
+    @property
+    def exchange_strs(self) -> List[str]:
+        return sorted(self.exchs_dict.keys())
+
+    @property
+    def n_input_feeds(self) -> int:
+        return len(self.input_feeds)
+
+    @property
+    def input_feeds(self) -> ArgFeeds:
```

Comment:
This should not be "input_feeds" for all ss's; that label is only appropriate for ai_model_ss. I would recommend to call it simply "feeds". 

Similar with "predict feed" for single feed ss. "Predict" is only appropriate for ai_model_ss. So just call it "feed".